en/ros2/px4_sdk.md

```diff
@@ -0,0 +1,336 @@
+- Integrated with the failsafe state machine and arming checks.
+- It is well-defined which setpoint types can be sent.
+- It is possible to replace an FMU-internal mode (such as RTL).
+
+## Getting Started
+
+The following steps are required to get started:
+
+1. Make sure you have a working [ROS 2 setup](../ros/ros2_comm.md), with _px4_msgs_ in the ROS 2 workspace.
+2. Clone the SDK into the workspace:
+
+   ```shell
+   cd $ros_workspace/src
+   git clone --recursive https://github.com/PX4/px4_sdk.git
+   ```
+
+   :::note
+   To ensure compatibility, use the latest _main_ branches for PX4, _px4_msgs_ and the SDK.
+   See also [here](https://github.com/PX4/px4_sdk#compatibility-with-px4).
+   :::
+
+3. Build the workspace:
+
+   ```shell
+   cd ..
+   source install/setup.bash
+   colcon build
+   ```
+
+4. In a different shell, start PX4 SITL (you can use any model or simulator):
+
+   ```shell
+   cd $px4-autopilot
+   make px4_sitl gazebo-classic
+   ```
+
+5. Run the micro XRCE agent in a new shell (you can keep it running afterward):
+
+   ```shell
+   MicroXRCEAgent udp4 -p 8888
+   ```
+
+6. Start QGroundControl.
+
+   :::note
+   Use QGroundControl Daily, which supports dynamically updating the list of modes.
+   :::
+
+7. Back in the ROS 2 terminal, run one of the example modes:
+
+   ```shell
+   ros2 run example_mode_manual_cpp example_mode_manual
+   ```
+
+   You should get an output like this:
+
+   ```
+   [DEBUG] [example_mode_manual]: Checking message compatibility...
+   [DEBUG] [example_mode_manual]: Subscriber found, continuing
+   [DEBUG] [example_mode_manual]: Publisher found, continuing
+   [DEBUG] [example_mode_manual]: Registering 'My Manual Mode' (arming check: 1, mode: 1, mode executor: 0)
+   [DEBUG] [example_mode_manual]: Subscriber found, continuing
+   [DEBUG] [example_mode_manual]: Publisher found, continuing
+   [DEBUG] [example_mode_manual]: Got RegisterExtComponentReply
+   [DEBUG] [example_mode_manual]: Arming check request (id=1, only printed once)
+   ```
+
+8. On the PX4 shell, you can check that PX4 registered the new mode:
+
+   ```shell
+   commander status
+   ```
+
+   The output should contain:
+
+   ```{5}
+   INFO  [commander] Disarmed
+   INFO  [commander] navigation mode: Position
+   INFO  [commander] user intended navigation mode: Position
+   INFO  [commander] in failsafe: no
+   INFO  [commander] External Mode 1: nav_state: 23, name: My Manual Mode
+   ```
+
+9. At this point you should be able to see the mode in QGroundControl as well:
+
+   ![QGC Modes](../../assets/middleware/ros2/px4_sdk/qgc_modes.png)
+
+10. Select the mode, make sure you have a manual control source (physical or virtual joystick), and arm the vehicle.
+    The mode will then activate, and it should print the following output:
+
+    ```
+    [DEBUG] [example_mode_manual]: Mode 'My Manual Mode' activated
+    ```
+
+11. Now you are ready to create your own mode.
+
+## Mode Tutorial
```

Comment:
You could call this a tutorial, but I'd called it a partially worked example. I think this is very useful, I'm just not sure that there isn't "more to it". For example, how do we associate this with a node? Why is it in an hpp file/does it matter.

This would be better as part of "How to use the SDK" - say as "Mode Class Definition", and separately have a full tutorial that works through the real example more thoroughly.

Reply:
PS There is obviously some magic that happens to tell PX4 it has a new class. It might be worth identifying that part - could be as simple as "Deriving from XX registers the mode using internal APIs" or similar. That would give hints to someone who wants to understand the underlying code, or if a class didn't register properly.